libraries/networking/src/ByteSlice.cpp

```diff
@@ -0,0 +1,96 @@
+//
+//  Byteslice.cpp
+//  libraries/networking/src
+//
+//  Created by Heather Anderson on 2020-05-01.
+//  Copyright 2020 Vircadia.
+//
+//  Distributed under the Apache License, Version 2.0.
+//  See the accompanying file LICENSE or http://www.apache.org/licenses/LICENSE-2.0.html
+//
+
+#include "Byteslice.h"
+
+quint8 ByteSlice::_fallback{ 0 };
+
+ByteSlice::ByteString::ByteString(const quint8* source, size_t length) {
+    quint8* myContent = new quint8[length];
+    memcpy(myContent, source, length);
+    _content = myContent;
+    _fullLength = length;
+}
+
+ByteSlice::ByteString::ByteString(size_t length) {
+    quint8* myContent = new quint8[length];
+    memset(myContent, 0, length);
+    _content = myContent;
+    _fullLength = length;
+}
+
+ByteSlice::ByteString::~ByteString() {
+    delete[] _content;
+}
+
+void ByteSlice::clear() {
+    _offset = 0;
+    _length = 0;
+    _content.reset();
+}
+
+quint8 ByteSlice::pop_front() {
+    if (!_length) {
+        return 0;
+    }
+    --_length;
+    return _content->_content[_offset++];
+}
+
+ByteSlice ByteSlice::substring(size_t offset, size_t length) const {
+    if (offset >= _length) {
+        return ByteSlice();
+    }
+    size_t maxLen = _length - offset;
+    if (length == NPOS) {
+        return ByteSlice(_content, _offset + offset, maxLen);
+    } else {
+        return ByteSlice(_content, _offset + offset, std::min(maxLen, length));
+    }
+}
+
+// create a new buffer and return a pointer to it
+void* ByteSlice::create(size_t length) {
+    _offset = 0;
+    _length = length;
+    _content = ByteStringPointer::create(length);
+    return const_cast<quint8*>(_content->_content);
```

Comment:
this method returns a void* so this cast isn’t necessary

Reply:
1) The compiler diagrees:
```
10>C:\devel\vircadia-udt3\libraries\networking\src\ByteSlice.cpp(65,30): error C2440: 'return': cannot convert from 'const quint8 *' to 'void *'
10>C:\devel\vircadia-udt3\libraries\networking\src\ByteSlice.cpp(65,20): message : Conversion loses qualifiers
```

2) This function is the *only* place that a non-const pointer can be received from a ByteString (i.e. the const_cast is very intentional here).  Every other member outside of create() would return an immutable pointer.

Reply:
ah!  my bad, I had misread this, you’re totally right